server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -114,6 +114,7 @@ public class IndexShardRoutingTable {
                 allShardsStarted = false;
             }
         }
+        assert primary != null || shards.isEmpty() : shards;
```

Comment:
I am surprised we allow empty case here. When is that possible?

Reply:
Unfortunately empty routing tables are pervasive in tests.